controllers/envoy/envoyconfigrevision_controller.go

```diff
@@ -79,44 +67,60 @@ func (r *EnvoyConfigRevisionReconciler) Reconcile(req ctrl.Request) (ctrl.Result
+			// Only the published version deletes the nodeID from the cache
+			if ecr.Status.Conditions.IsTrueFor(envoyv1alpha1.RevisionPublishedCondition) {
+				r.finalizeEnvoyConfigRevision(ecr.Spec.NodeID)
+				r.Log.Info("Successfully cleared xDS server cache", "XDSS", string(ecr.GetEnvoyAPIVersion()), "NodeID", ecr.Spec.NodeID)
+			}
+			// Remove EnvoyConfigFinalizer. Once all finalizers have been
+			// removed, the object will be deleted.
+			controllerutil.RemoveFinalizer(ecr, envoyv1alpha1.EnvoyConfigRevisionFinalizer)
+			if err := r.Client.Update(ctx, ecr); err != nil {
 				return ctrl.Result{}, err
 			}
-			// This is an unrecoverable error because resources are wrong
-			// so do not reque
-			return ctrl.Result{}, nil
 		}
+		return ctrl.Result{}, nil
+	}
 
-		oldSnap, err := r.XdsCache.GetSnapshot(nodeID)
-		// Publish the generated snapshot when the version is different from the published one. We look specifically
-		// for the version of the "Secret" resources because secrets can change even when the spec hasn't changed.
-		// Publish the snapshot when an error retrieving the published one occurs as it means that no snpshot has already
-		// been written to the cache for that specific nodeID.
-		if snap.GetVersion(envoy_resources.Secret) != oldSnap.GetVersion(envoy_resources.Secret) || err != nil {
-			r.Log.Info("Publishing new snapshot for nodeID", "Version", version, "NodeID", nodeID)
-			r.Log.V(1).Info("Writting to xDS cache", "CurrentVersion", oldSnap.GetVersion(envoy_resources.Secret),
-				"NewVersion", snap.GetVersion(envoy_resources.Secret))
-			if err := r.XdsCache.SetSnapshot(nodeID, snap); err != nil {
-				return ctrl.Result{}, err
+	// If this ecr has the RevisionPublishedCondition set to "True" pusblish the resources
+	// to the xds server cache
+	if ecr.Status.Conditions.IsTrueFor(envoyv1alpha1.RevisionPublishedCondition) {
+		decoder := envoy_serializer.NewResourceUnmarshaller(ecr.GetSerialization(), r.APIVersion)
+
+		cacheReconciler := reconcilers_envoy.NewCacheReconciler(
+			ctx, r.Log, r.Client, r.XdsCache,
+			decoder,
+			envoy_resources.NewGenerator(r.APIVersion),
+		)
+
+		result, err := cacheReconciler.Reconcile(req.NamespacedName, ecr.Spec.EnvoyResources, ecr.Spec.NodeID, ecr.Spec.Version)
+
+		// If a type errors.StatusError is returned it means that the config in spec.envoyResources is wrong
+		// and cannot be written into the xDS cache. This is true for any error loading all types of resources
+		// except for Secrets. Secrets are dynamically loaded from the API and transient failures are possible, so
+		// setting a permanent taint could occur for a transient failure, which is not desirable.
+		if result.Requeue || err != nil {
+			switch err.(type) {
+			case *errors.StatusError:
+				if err := r.taintSelf(ctx, ecr, "FailedLoadingResources", err.Error()); err != nil {
+					return ctrl.Result{}, err
+				}
+			default:
+				return result, err
```

Comment:
I realised that marking an EnvoyConfigRevision as `tainted` when errors occur while reading certificates from secrets is a problem, as transitory problems connecting the API server could end up tainting a perfectly valid EnvoyConfigRevision.